# ПЗ 1. Алгоритм линейной регрессии. Градиентный спуск.
### Практическое задание <a class="anchor" id="hw"></a><center>

In [1]:
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d.axes3d import Axes3D

1. Подберите скорость обучения (eta) и количество итераций

In [2]:
X = np.array([[ 1,  1],
              [ 1,  1],
              [ 1,  2],
              [ 1,  5],
              [ 1,  3],
              [ 1,  0],
              [ 1,  5],
              [ 1, 10],
              [ 1,  1],
              [ 1,  2]])

In [3]:
X.shape

(10, 2)

In [4]:
y = [45, 55, 50, 55, 60, 35, 75, 80, 50, 60]

In [7]:
def calc_mse(y, y_pred):
    err = np.mean(np.abs(y - y_pred))
    return err

In [16]:
n = X.shape[1]
err = None

eta = 1e-2 
n_iter = 40

W = np.array([1, 0.5])
print(f'Number of objects = {n} \
       \nLearning rate = {eta} \
       \nInitial weights = {W} \n')

for i in range(n_iter):
    y_pred = np.dot(X, W)
    err = calc_mse(y, y_pred)
    for k in range(W.shape[0]):
        W[k] -= eta * (1/n * 2 * X[:, k] @ (y_pred - y))
    if i % 10 == 0:
        eta /= 1.1
        print(f'Iteration #{i}: W_new = {W}, MSE = {round(err, 2)}')

Number of objects = 2        
Learning rate = 0.01        
Initial weights = [1.  0.5] 

Iteration #0: W_new = [ 6.4  19.35], MSE = 54.0
Iteration #10: W_new = [18.77319138  8.6270072 ], MSE = 18.93
Iteration #20: W_new = [27.15289315  7.06014716], MSE = 13.63
Iteration #30: W_new = [32.4241672   6.10426679], MSE = 10.31


***MSE = 10.31 при n_iter = 0.01 и количестве итераций 40***

2*. В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее

In [28]:
n = X.shape[1]

eta = 1e-2 
n_iter = 100

W = np.array([1, 0.5])
print(f'Number of objects = {n} \
       \nLearning rate = {eta} \
       \nInitial weights = {W} \n')

for i in range(n_iter):
    y_pred = np.dot(X, W)
    err = calc_mse(y, y_pred)
    
    # ИЗМЕНЕНИЯ
    W -= eta * (1/n * 2 * np.dot(X[:, 1], y_pred - y))
    # ИЗМЕНЕНИЯ

    if i % 10 == 0:
        print(f'Iteration #{i}: W_new = {W}, MSE = {round(err,2)}')

Number of objects = 2        
Learning rate = 0.01        
Initial weights = [1.  0.5] 

Iteration #0: W_new = [19.85 19.35], MSE = 54.0
Iteration #10: W_new = [19.85 19.35], MSE = 54.0
Iteration #20: W_new = [19.85 19.35], MSE = 54.0
Iteration #30: W_new = [19.85 19.35], MSE = 54.0
Iteration #40: W_new = [19.85 19.35], MSE = 54.0
Iteration #50: W_new = [19.85 19.35], MSE = 54.0
Iteration #60: W_new = [19.85 19.35], MSE = 54.0
Iteration #70: W_new = [19.85 19.35], MSE = 54.0
Iteration #80: W_new = [19.85 19.35], MSE = 54.0
Iteration #90: W_new = [19.85 19.35], MSE = 54.0
